File: openbis_api/source/java/ch/ethz/sis/openbis/generic/shared/api/v3/dto/search/DateFieldSearchCriterion.java
/*
 * Copyright 2014 ETH Zuerich, CISD
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package ch.ethz.sis.openbis.generic.shared.api.v3.dto.search;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

import ch.systemsx.cisd.base.annotation.JsonObject;

@JsonObject("DateFieldSearchCriterion")
public class DateFieldSearchCriterion extends AbstractFieldSearchCriterion<IDate>
{

    private static final long serialVersionUID = 1L;

    private ITimeZone timeZone = new ServerTimeZone();

    private IDateFormat dateFormat = new NormalDateFormat();

    DateFieldSearchCriterion(String fieldName, SearchFieldType fieldType)
    {
        super(fieldName, fieldType);
    }

    public void thatEquals(Date date)
    {
        setFieldValue(new DateObjectEqualToValue(date));
    }

    public void thatEquals(String date)
    {
        setFieldValue(new DateEqualToValue(date));
    }

    public void thatIsLaterThanOrEqualTo(Date date)
    {
        setFieldValue(new DateObjectLaterThanOrEqualToValue(date));
    }

    public void thatIsLaterThanOrEqualTo(String date)
    {
        setFieldValue(new DateLaterThanOrEqualToValue(date));
    }

    public void thatIsEarlierThanOrEqualTo(Date date)
    {
        setFieldValue(new DateObjectEarlierThanOrEqualToValue(date));
    }

    public void thatIsEarlierThanOrEqualTo(String date)
    {
        setFieldValue(new DateEarlierThanOrEqualToValue(date));
    }

    public DateFieldSearchCriterion withServerTimeZone()
    {
        this.timeZone = new ServerTimeZone();
        return this;
    }

    public DateFieldSearchCriterion withTimeZone(int hourOffset)
    {
        this.timeZone = new TimeZone(hourOffset);
        return this;
    }

    public DateFieldSearchCriterion withShortFormat()
    {
        this.dateFormat = new ShortDateFormat();
        return this;
    }

    public DateFieldSearchCriterion withNormalFormat()
    {
        this.dateFormat = new NormalDateFormat();
        return this;
    }

    public DateFieldSearchCriterion withLongFormat()
    {
        this.dateFormat = new LongDateFormat();
        return this;
    }

    public void setDateFormat(IDateFormat dateFormat)
    {
        checkValueFormat(getFieldValue(), dateFormat);
        this.dateFormat = dateFormat;
    }

    public IDateFormat getDateFormat()
    {
        return dateFormat;
    }

    public void setTimeZone(ITimeZone timeZone)
    {
        this.timeZone = timeZone;
    }

    public ITimeZone getTimeZone()
    {
        return timeZone;
    }

    @Override
    public void setFieldValue(IDate value)
    {
        checkValueFormat(value, getDateFormat());
        super.setFieldValue(value);
    }

    private static void checkValueFormat(IDate value, IDateFormat format)
    {
        if (value instanceof AbstractDateValue)
        {
            try
            {
                SimpleDateFormat simpleDateFormat = new SimpleDateFormat(format.getFormat());
                simpleDateFormat.setLenient(false);
                simpleDateFormat.parse(((AbstractDateValue) value).getValue());
            } catch (ParseException e)
            {
                throw new IllegalArgumentException("Date value: " + value + " does not match the format: " + format);
            }
        }
    }

}

File: js-test/servers/common/core-plugins/tests/1/as/webapps/openbis-v3-api-test/html/models/dto/search/DateFieldSearchCriterion.js
define([ "support/stjs", "sys/exceptions", "sys/simpledateformat", "dto/search/AbstractFieldSearchCriterion", "dto/search/ServerTimeZone", "dto/search/DateObjectEqualToValue",
		"dto/search/DateEqualToValue", "dto/search/DateObjectLaterThanOrEqualToValue", "dto/search/DateLaterThanOrEqualToValue", "dto/search/DateObjectEarlierThanOrEqualToValue",
		"dto/search/DateEarlierThanOrEqualToValue", "dto/search/TimeZone", "dto/search/AbstractDateValue", "dto/search/ShortDateFormat", "dto/search/NormalDateFormat", "dto/search/LongDateFormat" ],
		function(stjs, exceptions, SimpleDateFormat, AbstractFieldSearchCriterion, ServerTimeZone, DateObjectEqualToValue, DateEqualToValue, DateObjectLaterThanOrEqualToValue,
				DateLaterThanOrEqualToValue, DateObjectEarlierThanOrEqualToValue, DateEarlierThanOrEqualToValue, TimeZone, AbstractDateValue, ShortDateFormat, NormalDateFormat, LongDateFormat) {
			var DateFieldSearchCriterion = function(fieldName, fieldType) {
				AbstractFieldSearchCriterion.call(this, fieldName, fieldType);
				this.timeZone = new ServerTimeZone();
			};

			stjs.extend(DateFieldSearchCriterion, AbstractFieldSearchCriterion, [ AbstractFieldSearchCriterion ], function(constructor, prototype) {
				prototype['@type'] = 'dto.search.DateFieldSearchCriterion';
				constructor.serialVersionUID = 1;
				constructor.DATE_FORMATS = [ new ShortDateFormat(), new NormalDateFormat(), new LongDateFormat() ];
				prototype.thatEquals = function(date) {
					this.setFieldValue(new DateObjectEqualToValue(date));
				};
				prototype.thatEquals = function(date) {
					this.setFieldValue(new DateEqualToValue(date));
				};
				prototype.thatIsLaterThanOrEqualTo = function(date) {
					this.setFieldValue(new DateObjectLaterThanOrEqualToValue(date));
				};
				prototype.thatIsLaterThanOrEqualTo = function(date) {
					this.setFieldValue(new DateLaterThanOrEqualToValue(date));
				};
				prototype.thatIsEarlierThanOrEqualTo = function(date) {
					this.setFieldValue(new DateObjectEarlierThanOrEqualToValue(date));
				};
				prototype.thatIsEarlierThanOrEqualTo = function(date) {
					this.setFieldValue(new DateEarlierThanOrEqualToValue(date));
				};
				prototype.withServerTimeZone = function() {
					this.timeZone = new ServerTimeZone();
					return this;
				};
				prototype.withTimeZone = function(hourOffset) {
					this.timeZone = new TimeZone(hourOffset);
					return this;
				};
				prototype.setTimeZone = function(timeZone) {
					this.timeZone = timeZone;
				};
				prototype.getTimeZone = function() {
					return this.timeZone;
				};
				prototype.setFieldValue = function(value) {
					DateFieldSearchCriterion.checkValueFormat(value);
					AbstractFieldSearchCriterion.prototype.setFieldValue.call(this, value);
				};
				constructor.checkValueFormat = function(value) {
					if (stjs.isInstanceOf(value.constructor, AbstractDateValue)) {
						for ( var dateFormat in DateFieldSearchCriterion.DATE_FORMATS.getItems()) {
							try {
								var simpleDateFormat = new SimpleDateFormat(dateFormat.getFormat());
								simpleDateFormat.setLenient(false);
								simpleDateFormat.parse((value).getValue());
								return;
							} catch (e) {
							}
						}
						throw new exceptions.IllegalArgumentException("Date value: " + value + " does not match any of the supported formats: " + DateFieldSearchCriterion.DATE_FORMATS);
					}
				};
			}, {
				DATE_FORMATS : {
					name : "List",
					arguments : [ "IDateFormat" ]
				},
				timeZone : "ITimeZone",
				fieldType : {
					name : "Enum",
					arguments : [ "SearchFieldType" ]
				}
			});

			return DateFieldSearchCriterion;
		})
